aten/src/ATen/native/CompositeRandomAccessorCommon.h

```diff
@@ -0,0 +1,258 @@
+    : keys(keys), values(values)
+  {}
+
+  // Pointer-like operations {
+  C10_HOST_DEVICE
+  reference operator*() {
+    return TupleInfo::tie(*keys, *values);
+  }
+
+  // operator->() is supposed to return a pointer type.
+  // Since CompositeRandomAccessor does not hold pointers to pairs,
+  // we just return a pointer to a key.
+  C10_HOST_DEVICE
+  auto* operator->() const {
+    return keys.operator->();
+  }
+
+  C10_HOST_DEVICE
+  reference operator[](difference_type idx) {
+    return operator_brackets_proxy<self_type>(
+      CompositeRandomAccessor(keys + idx, values + idx)
+    );
+  }
+  // }
+
+  // Prefix/postfix increment/decrement {
+  C10_HOST_DEVICE
+  CompositeRandomAccessor& operator++() {
+    ++keys;
+    ++values;
+    return *this;
+  }
+
+  C10_HOST_DEVICE
+  CompositeRandomAccessor operator++(int) {
+    CompositeRandomAccessor copy(*this);
+    ++*this;
+    return copy;
+  }
+
+  C10_HOST_DEVICE
+  CompositeRandomAccessor& operator--() {
+    --keys;
+    --values;
+    return *this;
+  }
+
+  C10_HOST_DEVICE
+  CompositeRandomAccessor operator--(int) {
+    CompositeRandomAccessor copy(*this);
+    --*this;
+    return copy;
+  }
+  // }
+
+  // Arithmetic operations {
+  C10_HOST_DEVICE
+  CompositeRandomAccessor& operator+=(difference_type offset) {
+    keys += offset;
+    values += offset;
+    return *this;
+  }
+
+  C10_HOST_DEVICE
+  CompositeRandomAccessor operator+(difference_type offset) const {
+    return CompositeRandomAccessor(keys + offset, values + offset);
+  }
+
+  C10_HOST_DEVICE
+  friend CompositeRandomAccessor operator+(
+    difference_type offset,
+    const CompositeRandomAccessor& accessor
+  ) {
+    return accessor + offset;
+  }
+
+  C10_HOST_DEVICE
+  CompositeRandomAccessor& operator-=(difference_type offset) {
+    keys -= offset;
+    values -= offset;
+    return *this;
+  }
+
+  C10_HOST_DEVICE
+  CompositeRandomAccessor operator-(difference_type offset) const {
+    return CompositeRandomAccessor(keys - offset, values - offset);
+  }
+
+  C10_HOST_DEVICE
+  difference_type operator-(const CompositeRandomAccessor& other) const {
+    return keys - other.keys;
+  }
+  // }
+
+  // Comparison operators {
+  C10_HOST_DEVICE
+  bool operator==(const CompositeRandomAccessor& other) const {
+    return keys == other.keys;
+  }
+
+  C10_HOST_DEVICE
+  bool operator!=(const CompositeRandomAccessor& other) const {
+    return keys != other.keys;
+  }
+
+  C10_HOST_DEVICE
+  bool operator<(const CompositeRandomAccessor& other) const {
+    return keys < other.keys;
+  }
+
+  C10_HOST_DEVICE
+  bool operator<=(const CompositeRandomAccessor& other) const {
+    return keys <= other.keys;
+  }
+
+  C10_HOST_DEVICE
+  bool operator>(const CompositeRandomAccessor& other) const {
+    return keys > other.keys;
+  }
+
+  C10_HOST_DEVICE
+  bool operator>=(const CompositeRandomAccessor& other) const {
+    return keys >= other.keys;
+  }
+  // }
+
+protected:
+  KeyAccessor keys;
```

Comment:
This confused me for a while.
I've thought we have a list of keyAccessors and a list of ValueAccessors.
Can we add some comment here to make this more clear?
And a long term suggestion (not necessary for this PR), change the KeyAccessor and ValueAccessor to "a list of", it could be much useful if we can compare multiple layer values (eg, we have multiple input tensors and we need to compare them in order)

Reply:
The key is unique, otherwise `typename KeyAccessor::pointer` is not defined. `ValueAccessor` could be a composite type, that is true.

Reply:
Yes, not for this implementation.  
I basically want to see whether this can be extended to sort a tensor based on multiple input tensors. A list of ValueAccessors is not composite, it is just a list of ValueAccessor for different tensors, there might be multiple output tensors that need to write values into (other than indices). Nvm here, just some thoughts.

Reply:
Yeah, I am not sure that `keys` could be `list of`, because `KeyAcessor::pointer` type is not then defined, then the `CompositeRandomAccessor` becomes not a Random Access Iterator (note that `typename CompositeRadomAccessor::pointer = typename KeyAccessor::pointer`), so I am not sure STL will produce correct results in general.

Reply:
@nikitaved Nvm here. To achieve a list of KeyAcessor, the KeyAccessor itself might need some change, eg. hold a list of ptrs and strides, etc. Not for this PR.

Reply:
@glaringlee , for some reason I cannot comment under your message below ;) Could you please give an example what you mean by sorting a tensor based on multiple inputs? Multiple outputs should not be a problem, because `ValueAccessor` could be `CompositeRandomAccessor`.

Reply:
I see, then as long as `KeyAccessor` is a Random Access Iterator, it should work.